### ISSUE TITLE
enhancement: schema disagreement metric

### ISSUE THREAD

**Issue by avikivity** [scylladb/scylladb#11771]
It's often difficult to understand if a schema mismatch is eventually resolved, or if it stays forever. This is because the client may give up and go away, and we are left with collecting the logs and analyzing them.

To improve this, we can have a schema disagreement metric, nominally at zero and non-zero when a schema disagreement is detected. This allows us to understand at a glance is a schema disagreement was resolved and how long it took. It leverages the fact that metric collection and collation is easier and more reliable than the same thing with logs.

**Comment by avikivity** [comment]
@tgrabiec @asias please suggest implementations

@eliransin suggested count-of-nodes-that-disagree-with-my-schema

**Comment by asias** [comment]
What about each node publishes the schema UUID directly as a metric. If UUID is not possible, we can convert them to integers (e.g., 128bits or 64bits + 64bits).

With this, we can know exactly which node is out of sync over time.

**Comment by avikivity** [comment]
We can also export it as a label. But why won't a disagreement metric work? The disagreeing node will have N-1 disagreements, the others will have 1.

**Comment by avikivity** [comment]
(I like metrics to refer to a count of something, not be completely arbitrary)

**Comment by asias** [comment]
> We can also export it as a label. But why won't a disagreement metric work? The disagreeing node will have N-1 disagreements, the others will have 1.

Exporting the schema id has less indirection. It is the source of the true. Say n1 sees 2 nodes have different schema. We do not which node nodes are out of sync.

**Comment by avikivity** [comment]
> > We can also export it as a label. But why won't a disagreement metric work? The disagreeing node will have N-1 disagreements, the others will have 1.
> 
> Exporting the schema id has less indirection. It is the source of the true. Say n1 sees 2 nodes have different schema. We do not which node nodes are out of sync.

That's true. I just don't a metric that has a random value. We'd also need 4 ints, because the tools might convert int64->double along the way and lose precision.

**Comment by tgrabiec** [comment]
With raft, schema version could be (will be?) a timestamp, then by exporting it as a metric we would immediately know which node is behind and which is ahead.

What to export depends on what you really want. If you just want to see how long the disagreement took, then exporting a boolean called "in disagreement" is enough.
